File: backend/app/schemas/collectables.py
from pydantic import BaseModel


class GoalOut(BaseModel):
    type: str
    threshold: int
    current: int


class DropOut(BaseModel):
    slug: str
    name: str
    emoji: str
    type: str
    rarity: str | None
    ends_at: str | None
    goal: GoalOut
    earned: bool


class OwnedOut(BaseModel):
    slug: str
    name: str
    emoji: str
    type: str
    rarity: str | None
    equipped: bool


class CollectablesResponse(BaseModel):
    active: list[DropOut]
    owned: list[OwnedOut]

File: frontend/src/api/collectables.ts
import { useQuery } from '@tanstack/react-query';
import { apiFetch } from './client';

export type CollectableGoal = { type: string; threshold: number; current: number };

export type CollectableDrop = {
  slug: string;
  name: string;
  emoji: string;
  type: string;
  rarity: string | null;
  ends_at: string | null;
  goal: CollectableGoal;
  earned: boolean;
};

export type OwnedCollectable = {
  slug: string;
  name: string;
  emoji: string;
  type: string;
  rarity: string | null;
  equipped: boolean;
};

export type CollectablesState = { active: CollectableDrop[]; owned: OwnedCollectable[] };

export const getCollectables = () => apiFetch<CollectablesState>('/collectables');

export function useCollectables() {
  return useQuery({ queryKey: ['collectables'], queryFn: getCollectables });
}
